### DISCUSSION TITLE
Proposal: URI feature syntax

### DISCUSSION THREAD

**Discussion by mbasaglia** [lottie/lottie-spec#3] [Ideas]
We need a way of referencing objects, properties, etc within the specs.  This is useful in several places, for example profile definitions (#1) and [can-i-lottie](https://github.com/lottie-animation-community/caniuse/).

Assuming the schema in #2 is merged, I propose using URI-references similar to what `$ref` would use in the schema to refer to a specific definition.

Examples:
```
#/$defs/shapes/rectangle
#/$defs/shapes/rectangle/position
```

These are technically relative to the schema but I don't think they need to be fully qualified.

For profile definitions I propose using the notation described above to include parts of the schema, and the `!` prefix to exclude specific definitions.

**Comment by mbasaglia** [comment]
One thing I'm not sure of is the initial `#`, it's what JSON schema does but maybe it's better without? `$defs/shapes/rectangle` is still a valid relative URI reference.

**Comment by b-wils** [comment]
I'm wondering if we should support a more verbose profile definition. I could see having additional info such as the version a player added feature support or any caveats around a feature could be useful to encode somewhere.

Also not sure how tightly we want to couple profile syntax to JSON schema syntax. It probably makes any tooling we want to build more straightforward but may be out of place where there is no schema processing.

**Reply by mbasaglia** [reply]
> I could see having additional info such as the version a player added feature support or any caveats around a feature could be useful to encode somewhere

That's what canilottie is for.

> Also not sure how tightly we want to couple profile syntax to JSON schema syntax. It probably makes any tooling we want to build more straightforward but may be out of place where there is no schema processing.

My thought was to have the ability to export a JSON schema for for profiles so it can be used for validation

**Reply by b-wils** [reply]
> That's what canilottie is for.

My thought was to eventually have canilottie be generated from the profiles, that may be ambitious though.

> My thought was to have the ability to export a JSON schema for for profiles so it can be used for validation

We could still have the profile use a shorthand syntax (eg `shapes/rectangle`) and the profile schema generation can flesh out the reference as long as the feature naming is consistent.

**Reply by mbasaglia** [reply]
> My thought was to eventually have canilottie be generated from the profiles

Hm, not sure they should work that way. Different players / versions will have quirks that should be shown on canilottie but do not being in the profiles

**Reply by b-wils** [reply]
> Different players / versions will have quirks that should be shown on canilottie but do not being in the profiles

I think we could include those quirks as metadata in the profile, but I could see that being a big challenge to cover everything for both use cases. Especially if we need to specify something more than basic text for canilottie.

**Comment by b-wils** [comment]
For profiles, I'd like to propose a nested JSON structure with grouped features . Something like-

```
 {
    "shapes": {
        "rectangle": true,
        "ellipse": false,
        // more shapes
    },
    "effects": {
        "black-and-white": false,
        "color-levels": true,
        // more effects
    }
    // more feature groups
}
```

I think it's got good readability and gives flexibility to add features that may not be easy to represent with a single schema reference.
